android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/BaseSearchTelemetry.kt

```diff
@@ -21,20 +25,47 @@ import mozilla.components.support.base.facts.collect
 import mozilla.components.support.base.log.logger.Logger
 import mozilla.components.support.ktx.kotlinx.coroutines.flow.filterChanged
 import org.json.JSONObject
+import java.io.File
+
+internal const val COLLECTION_NAME = "search-telemetry-v2"
 
 /**
  * Main configuration and functionality for tracking ads / web searches with specific providers.
  */
-abstract class BaseSearchTelemetry {
+abstract class BaseSearchTelemetry(private val file: File) {
+    @VisibleForTesting
+    internal var providerList: List<SearchProviderModel>? = emptyList()
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    internal fun initializeProviderList(): Job {
+        val fetcher = SerpTelemetryFetcher(file, COLLECTION_NAME)
+
+        return CoroutineScope(Dispatchers.IO).launch {
+            providerList = fetcher.fetchSearchProviders()
+        }
+    }
 
     /**
-     * Install the web extensions that this functionality is based on and start listening for updates.
+     * Finds provider among list of providers that matches regex in url.
      */
-    abstract fun install(engine: Engine, store: BrowserStore)
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    fun getProviderForUrl(url: String): SearchProviderModel? {
+        return providerList?.find { provider ->
+            provider.searchPageRegexp.pattern.toRegex().containsMatchIn(url)
+        }
+    }
 
     @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
-    internal fun getProviderForUrl(url: String): SearchProviderModel? =
-        providerList.find { provider -> provider.regexp.containsMatchIn(url) }
+    internal fun waitUntilProviderListInitialized() {
+        if (providerList.isNullOrEmpty()) {
+            initializeProviderList()
+        }
+    }
```

Comment:
I don't think we need this function, and here are a couple reasons why I think so:

1. We never actually wait for anything. This is launched inside a coroutine, so `installWebExtension` will complete instead of waiting. Similarly, `initializeProviderList` launches a coroutine and completes, and we are not actually waiting for the result here.
2. Since the places that `getProviderForUrl` is called already have good handling for `null` results, we don't actually need to wait for anything. We can launch a coroutine to fetch the results, and then if it hasn't completed by the time we call `getProviderForUrl` we can just rely on the normal null branch flow.

Reply:
True. Lingered over from last updates. Thank you for pointing this out